IPv6 Shared-Gateway fixes

Billy99 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

Billy99 commented:
Trying to determine if failures are flakes or an actual issue with PR. Below is a list of recent failure runs (shard-network for both IPv4 shared gateway and IPv6 shared gateway):

**09:37 AM  8/19/20**

- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv4, IPv4, true, false)

> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [BeforeEach] should enforce policy based on Ports [Feature:NetworkPolicy] 
> - [Fail] [sig-network] Network [It] should resolve connrection reset issue #74839 [Slow] 
> 

- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)

> - [Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
> - [Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols 

**08:36 AM  8/19/20**

- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv4, IPv4, true, false)
- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)

> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should stop enforcing policies after they are deleted [Feature:NetworkPolicy] 


**12:11 AM  8/19/20**

- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)
> - [Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols 
- ovn-ci / e2e (shard-network, false, false, noHA, local, ipv4, IPv4, true, false)
> - [Fail] [sig-network] Network [It] should resolve connrection reset issue #74839 [Slow] 


**10:39 PM  8/18/20**

- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv4, IPv4, true, false)
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce policies to check ingress and egress policies can becontrolled independently based on PodSelector [Feature:NetworkPolicy] 
- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce multiple egress policies with egress allow-all policy taking precedence [Feature:NetworkPolicy] 
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce policies to check ingress and egress policies can be controlled independently based on PodSelector [Feature:NetworkPolicy]
> - [Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols 
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce egress policy allowing traffic to a server in a different namespace based on PodSelector and NamespaceSelector [Feature:NetworkPolicy] 

**09:00 PM  8/18/20**

-  ovn-ci / e2e (shard-network, false, true, HA, shared, ipv4, IPv4, true, false)
> -   [Fail] [sig-network] Network [It] should resolve connrection reset issue #74839 [Slow] 
> -   [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should stop enforcing policies after they are deleted [Feature:NetworkPolicy] 
-  ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)
> -   [Fail] [sig-network] Services [It] should be able to change the type from ExternalName to NodePort [Conformance] 
> -   [Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
> -   [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce policies to check ingress and egress policies can be controlled independently based on PodSelector [Feature:NetworkPolicy] 
> -   [Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols

danwinship commented:
So is the last commit still needed given #1627?

danwinship commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -370,26 +370,51 @@ func addDefaultConntrackRules(nodeName, gwBridge, gwIntf string, stopChan chan s
+	}
+	if config.IPv6Mode {
+		_, stderr, err = util.RunOVSOfctl("add-flow", gwBridge,
+			fmt.Sprintf("cookie=%s, priority=100, in_port=%s, ipv6, "+
```
doh! the fact that `ip` means IPv4 only is so non-obvious sometimes

danwinship commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -370,26 +370,51 @@ func addDefaultConntrackRules(nodeName, gwBridge, gwIntf string, stopChan chan s
+	if config.IPv4Mode {
+		_, stderr, err = util.RunOVSOfctl("add-flow", gwBridge,
+			fmt.Sprintf("cookie=%s, priority=50, in_port=%s, ip, "+
+				"actions=ct(zone=%d, table=1)", defaultOpenFlowCookie, ofportPhys, config.Default.ConntrackZone))
```
might be clearer to do

    if config.IPv4Mode {
            // pod-to-external flow
            // external-to-pod flow
    }
    if config.IPv6Mode {
            // pod-to-external flow
            // external-to-pod flow
    }

rather than

    // pod-to-external flow
    if config.IPv4Mode {
            ...
    }
    if config.IPv6Mode {
            ...
    }
    // external-to-pod flow
    if config.IPv4Mode {
            ...
    }
    if config.IPv6Mode {
            ...
    }

Billy99 replied:
Done.

Billy99 commented:
@danwinship No, but I was waiting on @girishmg to confirm the flood approach was acceptable before I remove the commit to write the mac_bindings.

Billy99 commented:
> So is the last commit still needed given #1627?

Removed the first commit, add IPv6 shared-gateway to CI, because it won't pass with just this PR. Also removed the last commit, add MAC_Binding entries for each node in each Gateway Router, in favor of #1627.

Billy99 commented:
@trozet @danwinship PTAL
Seems to be a lot of flakes lately on CI runs. This is the best run I've had in a while. I rearranged the logic per Dan's request and removed the other commits in favor of #1613 and #1627.

trozet commented:
> @trozet @danwinship PTAL
> Seems to be a lot of flakes lately on CI runs. This is the best run I've had in a while. I rearranged the logic per Dan's request and removed the other commits in favor of #1613 and #1627.

The flakes are around a new bug with OVN, we are now skipping that test case: 
https://github.com/ovn-org/ovn-kubernetes/pull/1625

Also, network policy flakes (which you hit): https://github.com/ovn-org/ovn-kubernetes/issues/1624

aojea commented:
These are the failures in the ipv6 jobs, 
```
Summarizing 9 Failures:

[Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce egress policy allowing traffic to a server in a different namespace based on PodSelector and NamespaceSelector [Feature:NetworkPolicy] 
/home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:1522

[Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols 
/home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/service.go:1570

[Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce policies to check ingress and egress policies can be controlled independently based on PodSelector [Feature:NetworkPolicy] 
/home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:1522

[Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce egress policy allowing traffic to a server in a different namespace based on PodSelector and NamespaceSelector [Feature:NetworkPolicy] 
/home/runner/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/network_policy.go:1522

[Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
```
can we retest to be sure those are unrelated to the PR?

Billy99 commented:
Trying to determine if failures are flakes or an actual issue with PR. Below is a list of recent failure runs with PR #1627 on master and commit in this PR to update IPv6 OvS conntrack flows:

**10:16 AM  08/20/20** - https://github.com/ovn-org/ovn-kubernetes/actions/runs/216816514
- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce egress policy allowing traffic to a server in a different namespace based on PodSelector and NamespaceSelector [Feature:NetworkPolicy] 
> - [Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
> - [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [It] should enforce multiple egress policies with egress allow-all policy taking 

**12:45 PM 08/20/20** - https://github.com/ovn-org/ovn-kubernetes/actions/runs/217008090
- ovn-ci / e2e (shard-network, false, true, HA, shared, ipv6, IPv6, false, true)
> - [Fail] [sig-network] Services [It] should be able to create a functioning NodePort service [Conformance] 
> - [Fail] [sig-network] Services [It] should be able to change the type from ExternalName to NodePort [Conformance] 
> - [Fail] [sig-network] Services [It] should be able to update service type to NodePort listening on same port number but different protocols 
- ovn-ci / e2e (shard-network, false, false, noHA, local, ipv4, IPv4, true, false)
> - Cancelled

trozet commented:
Network Policy isn't that big of a concern, we have flakes there even in ipv4. However, I dont see this fail in v4:

Services [It] should be able to create a functioning NodePort service

That makes me think something is wrong with the flows on breth0. Need to look into this.

danwinship commented:
The actual gateway fix looks correct. We could merge that immediately and continue to debug other CI issues in other PRs

Billy99 commented:
Just leave out #1613. correct?

trozet commented:
@Billy99 I think I see the problem in services we add openflow rules directly as the proto being used in the service, however in openflow for example with v6 you need to add tcp6, not tcp

Billy99 commented:
ACK, I'll take a look

trozet commented:
@Billy99 here https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_shared_intf.go#L40
